core/jvm/src/main/scala/scalaz/zio/RTS.scala

```diff
@@ -588,7 +593,7 @@ private object RTS {
               // At this point, all causes of interruption have been accumulated
               // in the fiber status and will be read during evaluation of this
               // action:
-              curIo = IO.terminate
+              curIo = IO.interrupt
```

Comment:
Hmmm, this suggests to me the solution is more complex than just introducing `selfInterrupt`.

The runtime system really wants a term in `IO` for interruption, e.g.:

```scala
class Interrupt(causes: List[Throwable], defects: List[Throwable] extends IO[Nothing, Nothing]
```

permitting a method:

```scala
object IO {
  ...
  def interrupt(causes: List[Throwable], defects: List[Throwable])
  ...
}
```

which will perform self-interruption.

Introducing this term could actually simplify some core logic because it's possible (though not certain) we could remove causes / defects from the fiber state with enough work.